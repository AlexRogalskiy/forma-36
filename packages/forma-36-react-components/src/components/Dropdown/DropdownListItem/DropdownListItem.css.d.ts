// This file is automatically generated by typings-for-css-modules.
// Please do not change this file!
export interface IDropdownListItemCss {
  'DropdownListItem': string;
  'DropdownListItem__toggle-button': string;
  'DropdownListItem__submenu-toggle': string;
  'DropdownListItem__toggle-button__inner-wrapper': string;
  'DropdownListItem--disabled': string;
  'DropdownListItem--active': string;
  'DropdownListItem--title': string;
}

export const locals: IDropdownListItemCss;
