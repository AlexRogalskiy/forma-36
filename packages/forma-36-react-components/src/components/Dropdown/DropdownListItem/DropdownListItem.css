@import 'settings/colors';
@import 'settings/typography';

.DropdownListItem {
  display: inline-block;
  padding: calc(1rem * (7 / var(--rem-base)))
    calc(1rem * (20 / var(--rem-base)));
  font-family: var(--font-stack-primary);
  font-size: var(--font-size-paragraph);
  white-space: nowrap;
  display: block;
  position: relative;
  color: var(--color-text-mid);
  font-weight: var(--font-weight-normal);
  margin: 0;
}

.DropdownListItem__toggle-button:hover {
  background: var(--color-element-lightest);
}

.DropdownListItem__submenu-toggle {
  padding: 0px;
}

.DropdownListItem__toggle-button {
  composes: focus-outline--default from './../../../styles/settings/a11y.css';
  border: none;
  background: transparent;
  font-family: var(--font-stack-primary);
  width: 100%;
  font-size: var(--font-size-paragraph);
  text-align: left;
  padding: 0;
  cursor: pointer;
  color: var(--color-text-mid);
}

.DropdownListItem__toggle-button__inner-wrapper {
  display: block;
  padding: calc(1rem * (7 / var(--rem-base)))
    calc(1rem * (20 / var(--rem-base)));
  line-height: 1.5;
}

.DropdownListItem--disabled,
.DropdownListItem--disabled button {
  background: var(--color-white);
  opacity: 0.5;
  cursor: not-allowed;

  &:hover {
    background: var(--color-white);
  }
}

.DropdownListItem--active,
.DropdownListItem--active:hover {
  background: var(--color-element-light);
}

.DropdownListItem--active .DropdownListItem__toggle-button:hover {
  background: var(--color-element-light);
  cursor: default;
}

.DropdownListItem--title {
  line-height: 1.8;
  padding-bottom: calc(1rem * (5 / var(--rem-base)));
}

.DropdownListItem--title,
.DropdownListItem--title:hover {
  color: var(--color-text-dark);
  background: var(--color-white);
  font-size: calc(1rem * (12 / var(--rem-base)));
  font-weight: var(--font-weight-demi-bold);
  text-transform: uppercase;
  letter-spacing: 0.1rem;
}
