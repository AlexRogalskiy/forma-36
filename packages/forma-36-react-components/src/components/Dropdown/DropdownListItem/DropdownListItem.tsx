import React, {
  Component,
  MouseEventHandler,
  FocusEventHandler,
  MouseEvent as ReactMouseEvent,
} from 'react';
import cn from 'classnames';
import TabFocusTrap from '../../TabFocusTrap/TabFocusTrap';
import styles from './DropdownListItem.css';

export type DropdownListItemProps = {
  isDisabled: boolean;
  isActive: boolean;
  isTitle: boolean;
  children: React.ReactNode;
  onClick?: MouseEventHandler;
  onMouseDown?: MouseEventHandler;
  submenuToggleLabel?: string;
  href?: string;
  onFocus?: FocusEventHandler;
  onLeave?: MouseEventHandler;
  onEnter?: MouseEventHandler;
  className?: string;
  testId?: string;
} & typeof defaultProps;

const defaultProps = {
  testId: 'cf-ui-dropdown-list-item',
  isDisabled: false,
  isActive: false,
  isTitle: false,
};

export class DropdownListItem extends Component<DropdownListItemProps> {
  static defaultProps = defaultProps;

  renderSubmenuToggle = () => {
    const { onClick, onEnter, onLeave, onFocus, ...otherProps } = this.props;
    return (
      <React.Fragment>
        <button
          type="button"
          data-test-id="cf-ui-dropdown-submenu-toggle"
          className={styles['DropdownListItem__button']}
          onClick={onClick}
          onMouseEnter={onEnter}
          onFocus={onFocus}
          onMouseLeave={onLeave}
          {...otherProps}
        >
          <TabFocusTrap
            className={styles['DropdownListItem__button__inner-wrapper']}
          >
            {this.props.submenuToggleLabel}
          </TabFocusTrap>
        </button>
        {this.props.children}
      </React.Fragment>
    );
  };

  renderListItem = () => {
    const isClickable =
      this.props.onClick || this.props.onMouseDown || this.props.href;

    if (isClickable) {
      const Element = this.props.href ? 'a' : 'button';

      return (
        <Element
          type="button"
          href={this.props.href}
          data-test-id="cf-ui-dropdown-list-item-button"
          className={styles['DropdownListItem__button']}
          onClick={(e: ReactMouseEvent) => {
            if (!this.props.isDisabled && this.props.onClick) {
              this.props.onClick(e);
            }
          }}
          onMouseDown={(e: ReactMouseEvent) => {
            if (!this.props.isDisabled && this.props.onMouseDown) {
              this.props.onMouseDown(e);
            }
          }}
        >
          <TabFocusTrap
            className={styles['DropdownListItem__button__inner-wrapper']}
          >
            {this.props.children}
          </TabFocusTrap>
        </Element>
      );
    }

    return this.props.children;
  };

  render() {
    const {
      className,
      isDisabled,
      testId,
      isActive,
      onClick,
      onMouseDown,
      href,
      submenuToggleLabel,
      isTitle,
    } = this.props;

    const classNames = cn(styles['DropdownListItem'], className, {
      [styles['DropdownListItem__submenu-toggle']]:
        submenuToggleLabel || onClick || onMouseDown || href,
      [styles['DropdownListItem--disabled']]: isDisabled,
      [styles['DropdownListItem--active']]: isActive,
      [styles['DropdownListItem--title']]: isTitle,
    });

    return (
      <li className={classNames} data-test-id={testId}>
        {submenuToggleLabel
          ? this.renderSubmenuToggle()
          : this.renderListItem()}
      </li>
    );
  }
}

export default DropdownListItem;
