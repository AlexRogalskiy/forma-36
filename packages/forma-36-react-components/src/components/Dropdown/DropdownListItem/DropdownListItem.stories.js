import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean } from '@storybook/addon-knobs';
import { host } from 'storybook-host';
import { withInfo } from '@storybook/addon-info';
import DropdownListItem from './DropdownListItem';

storiesOf('Components|Dropdown/DropdownListItem', module)
  .addDecorator(
    host({
      align: 'center middle',
      cropMarks: false,
    }),
  )
  .add(
    'default',
    withInfo()(() => (
      <DropdownListItem
        isDisabled={boolean('isDisabled', false)}
        isActive={boolean('isActive', false)}
        isTitle={boolean('isTitle', false)}
      >
        {text('Text', 'Menu Entry')}
      </DropdownListItem>
    )),
  );
