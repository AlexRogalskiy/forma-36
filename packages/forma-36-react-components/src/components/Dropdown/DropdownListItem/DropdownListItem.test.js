import React from 'react';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import DropdownListItem from './DropdownListItem';

it('renders the component', () => {
  const output = shallow(<DropdownListItem>DropdownMenuItem</DropdownListItem>);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <DropdownListItem extraClassNames="my-extra-class">
      <DropdownListItem>Entry</DropdownListItem>
    </DropdownListItem>,
  );

  expect(output).toMatchSnapshot();
});

it('renders the component as active', () => {
  const output = shallow(<DropdownListItem isActive>entry</DropdownListItem>);

  expect(output).toMatchSnapshot();
});

it('renders the component as disabled', () => {
  const output = shallow(<DropdownListItem isDisabled>entry</DropdownListItem>);

  expect(output).toMatchSnapshot();
});

it('renders the component as disabled', () => {
  const output = shallow(<DropdownListItem isDisabled>entry</DropdownListItem>);

  expect(output).toMatchSnapshot();
});

it('calls onClick', () => {
  const mockOnClick = sinon.spy();
  const dropDownItem = mount(
    <DropdownListItem onClick={mockOnClick}>entry</DropdownListItem>,
  );
  const button = dropDownItem.find('button');
  button.simulate('click');
  expect(mockOnClick.called).toBe(true);
});

it('does not call onClick on a disabled menu item', () => {
  const mockOnClick = sinon.spy();
  const dropDownItem = mount(
    <DropdownListItem isDisabled onClick={mockOnClick}>
      entry
    </DropdownListItem>,
  );
  const button = dropDownItem.find('button');
  button.simulate('click');
  expect(mockOnClick.called).toBe(false);
});
