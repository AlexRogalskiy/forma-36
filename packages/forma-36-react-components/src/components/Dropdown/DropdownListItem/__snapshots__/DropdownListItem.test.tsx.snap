// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<li
  class="DropdownListItem"
  data-test-id="cf-ui-dropdown-list-item"
  role="menuitem"
>
  <span>
    DropdownMenuItem
  </span>
</li>
`;

exports[`renders the component as active 1`] = `
<li
  class="DropdownListItem DropdownListItem--active"
  data-test-id="cf-ui-dropdown-list-item"
  role="menuitem"
>
  <span>
    entry
  </span>
</li>
`;

exports[`renders the component as disabled 1`] = `
<li
  class="DropdownListItem DropdownListItem--disabled"
  data-test-id="cf-ui-dropdown-list-item"
  role="menuitem"
>
  <span>
    entry
  </span>
</li>
`;

exports[`renders the component with a href 1`] = `
<li
  class="DropdownListItem DropdownListItem__submenu-toggle"
  data-test-id="cf-ui-dropdown-list-item"
  role="menuitem"
>
  <a
    class="DropdownListItem__button"
    data-test-id="cf-ui-dropdown-list-item-button"
    href="#"
    type="button"
  >
    <span
      class="TabFocusTrap DropdownListItem__button__inner-wrapper"
      tabindex="-1"
    >
      DropdownListItem with a href
    </span>
  </a>
</li>
`;

exports[`renders the component with an additional class name 1`] = `
<li
  class="DropdownListItem my-extra-class"
  data-test-id="cf-ui-dropdown-list-item"
  role="menuitem"
>
  <span
    class="my-extra-class"
  >
    Entry
  </span>
</li>
`;
